lib/ddtrace/contrib/active_support/notifications/subscriber.rb

```diff
@@ -36,8 +41,7 @@ def on_subscribe(&block)
             # Should be triggered by the inheriting class.
             def subscribe!
               return subscribed? if subscribed? || on_subscribe_block.nil?
-              on_subscribe_block.call
-              @subscribed = true
+              instance_eval(&on_subscribe_block)
```

Comment:
Why are we changing the scope of this? This is very likely to break user applications where they introduce blocks that are dependent on a local scope as opposed to the subscriber scope. Highly recommend reverting this.

Reply:
`instance_eval` only changes self, and allows accessing methods from class or class instance which otherwise wouldn't be. 
It however doesn't modify the binding of the block. 

So accessing methods and variables defined in 'binding' of the block will still be possible whether `block.call` or `instance_eval(&block)` is used.

To make your example work a user has to use two "protected" apis via `.send(method)` which in itself is accessing API's which shouldn't be accessible to our users.
i.e.

 ```ruby
Datadog::Contrib::ActiveRecord::Events::SQL.tap do |event|
  event.send(:on_subscribe) do
    event.send(:subscribe, event.event_name,  event.span_name, options, event.tracer)
  end
end
```
Note however this still works with current changes.

this change only allows APIs like to work as well as the version above, so its backwards compatible while still allowing nicer DSL. Broken test only showed that this DSLs api is quite fragile

 ```ruby
Datadog::Contrib::ActiveRecord::Events::SQL.send(:on_subscribe) do
  subscribe(event_name,  span_name, options, tracer)
end
```

Also note that this snippet will also work as expected.
 ```ruby
span_name = 'different_name'
Datadog::Contrib::ActiveRecord::Events::SQL.send(:on_subscribe) do
  subscribe(event_name,  span_name, options, tracer)
end
```

Overall, I don't mind leaving the scope as is since this API is 'protected' either way and I suspect only tests and our code uses it, so I can revert the change.